File: apps/backend/src/cli/tools/story.go
// apps/backend/src/cli/tools/story.go
package tools

import (
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/lucas/oraculo/apps/backend/src/db"
	"github.com/lucas/oraculo/apps/backend/src/domain"
	"github.com/lucas/oraculo/apps/backend/src/output"
	"github.com/spf13/cobra"
)

// newStoryCmd returns the "story" parent command with all subcommands.
func newStoryCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "story",
		Short: "Manage stories",
	}
	cmd.AddCommand(
		newStoryInitCmd(),
		newStorySaveCmd(),
		newStoryGetCmd(),
		newStoryListCmd(),
		newStoryUpdateCmd(),
		newStoryUpdateStatusCmd(),
		newStoryDeleteCmd(),
		newStoryVersionCmd(),
		newStoryVersionsCmd(),
	)
	return cmd
}

// resolveEpic looks up an epic by name and returns it. On error it writes
// the JSON error to w and returns nil plus the error.
func resolveEpic(cmd *cobra.Command) (*db.DB, string, int, error) {
	w := cmd.OutOrStdout()
	epicName, _ := cmd.Flags().GetString("epic")
	database := dbFromContext(cmd.Context())
	epicStore := db.NewEpicStore(database)
	epic, err := epicStore.GetByName(epicName)
	if err != nil {
		output.WriteError(w, err)
		return nil, "", 0, err
	}
	return database, epicName, epic.ID, nil
}

func newStoryInitCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "init <name>",
		Short: "Create a new story within an epic",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			description, _ := cmd.Flags().GetString("description")
			w := cmd.OutOrStdout()

			database, epicName, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			store := db.NewStoryStore(database)
			_, created, err := store.Create(epicID, name, description)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, map[string]any{
				"name":    name,
				"epic":    epicName,
				"path":    filepath.Join(".oraculo", "epics", epicName, "stories", name),
				"created": created,
			})
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	cmd.Flags().String("description", "", "Story description")
	return cmd
}

func newStorySaveCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "save <name>",
		Short: "Save requirements markdown for a story (reads stdin)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			database, epicName, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			// Upsert DB record so downstream commands (e.g. approval request) can find the story.
			store := db.NewStoryStore(database)
			if _, _, err := store.Create(epicID, name, ""); err != nil {
				output.WriteError(w, err)
				return err
			}

			content, err := io.ReadAll(cmd.InOrStdin())
			if err != nil {
				output.WriteError(w, fmt.Errorf("read stdin: %w", err))
				return err
			}

			dir := filepath.Join(".oraculo", "epics", epicName, "stories", name)
			if err := os.MkdirAll(dir, 0o755); err != nil {
				output.WriteError(w, fmt.Errorf("create directory: %w", err))
				return err
			}

			path := filepath.Join(dir, "requirements.md")
			if err := os.WriteFile(path, content, 0o644); err != nil {
				output.WriteError(w, fmt.Errorf("write file: %w", err))
				return err
			}

			return output.WriteJSON(w, map[string]any{
				"name":  name,
				"epic":  epicName,
				"path":  path,
				"saved": true,
			})
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

func newStoryGetCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "get <name>",
		Short: "Get requirements markdown for a story",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			_, epicName, _, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			path := filepath.Join(".oraculo", "epics", epicName, "stories", name, "requirements.md")
			content, err := os.ReadFile(path)
			if err != nil {
				wrappedErr := fmt.Errorf("story %q requirements: %w", name, err)
				output.WriteError(w, wrappedErr)
				return wrappedErr
			}

			_, err = w.Write(content)
			return err
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

func newStoryListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List all stories for an epic",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			w := cmd.OutOrStdout()

			database, _, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			store := db.NewStoryStore(database)
			stories, err := store.List(epicID)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, stories)
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

func newStoryUpdateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "update <name>",
		Short: "Update a story's description",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			description, _ := cmd.Flags().GetString("description")
			w := cmd.OutOrStdout()

			database, _, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			store := db.NewStoryStore(database)
			story, err := store.Update(epicID, name, description)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, story)
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	cmd.Flags().String("description", "", "New story description")
	return cmd
}

func newStoryUpdateStatusCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "update-status <name>",
		Short: "Update a story's approval status",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			statusStr, _ := cmd.Flags().GetString("status")
			w := cmd.OutOrStdout()

			status := domain.ApprovalStatus(statusStr)
			if !status.Valid() {
				err := fmt.Errorf("invalid status %q: must be one of none, pending, approved, rejected, needs_revision", statusStr)
				output.WriteError(w, err)
				return err
			}

			database, _, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			store := db.NewStoryStore(database)
			if err := store.UpdateApprovalStatus(epicID, name, status); err != nil {
				output.WriteError(w, err)
				return err
			}

			story, err := store.GetByName(epicID, name)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, story)
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	cmd.Flags().String("status", "", "Approval status: none|pending|approved|rejected|needs_revision (required)")
	_ = cmd.MarkFlagRequired("status")
	return cmd
}

func newStoryVersionCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "version <name>",
		Short: "Create a new version of story definition (reads stdin)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			database, _, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			storyStore := db.NewStoryStore(database)
			story, err := storyStore.GetByName(epicID, name)
			if err != nil {
				output.WriteError(w, err)
				return err
			}

			content, err := io.ReadAll(cmd.InOrStdin())
			if err != nil {
				output.WriteError(w, fmt.Errorf("read stdin: %w", err))
				return err
			}

			versionStore := db.NewVersionStore(database)
			version, err := versionStore.CreateStoryVersion(story.ID, string(content))
			if err != nil {
				output.WriteError(w, err)
				return err
			}

			approvalStore := db.NewApprovalStore(database)
			approval, err := approvalStore.Request(domain.ApprovalDesign, &epicID, &story.ID, string(content))
			if err != nil {
				output.WriteError(w, err)
				return err
			}

			return output.WriteJSON(w, map[string]any{
				"version_id":  version.ID,
				"approval_id": approval.ID,
			})
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

func newStoryVersionsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "versions <name>",
		Short: "List all versions of a story",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			database, _, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			storyStore := db.NewStoryStore(database)
			story, err := storyStore.GetByName(epicID, name)
			if err != nil {
				output.WriteError(w, err)
				return err
			}

			versionStore := db.NewVersionStore(database)
			versions, err := versionStore.ListStoryVersions(story.ID)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, versions)
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

func newStoryDeleteCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "delete <name>",
		Short: "Delete a story",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			database, epicName, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			store := db.NewStoryStore(database)
			if err := store.Delete(epicID, name); err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, map[string]any{
				"name":    name,
				"epic":    epicName,
				"deleted": true,
			})
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

File: apps/orchestrator/src/cli/tools/story.ts
import { Command } from "commander";
import { join } from "node:path";
import { mkdirSync } from "node:fs";
import { EpicStore } from "../../db/stores/epic-store.ts";
import { StoryStore } from "../../db/stores/story-store.ts";
import { VersionStore } from "../../db/stores/version-store.ts";
import { ApprovalStore } from "../../db/stores/approval-store.ts";
import { writeJSON, writeError } from "../../utils/output.ts";
import { APPROVAL_STATUSES } from "../../domain/enums.ts";
import type { ApprovalStatus } from "../../domain/enums.ts";
import { dbFromCmd } from "../context.ts";
import type { Database } from "bun:sqlite";

export function storyCommands(): Command {
  const cmd = new Command("story").description("Manage stories");

  cmd.addCommand(storyInitCmd());
  cmd.addCommand(storySaveCmd());
  cmd.addCommand(storyGetCmd());
  cmd.addCommand(storyListCmd());
  cmd.addCommand(storyUpdateCmd());
  cmd.addCommand(storyUpdateStatusCmd());
  cmd.addCommand(storyDeleteCmd());
  cmd.addCommand(storyVersionCmd());
  cmd.addCommand(storyVersionsCmd());

  return cmd;
}

function resolveEpic(cmd: Command): { db: Database; epicName: string; epicId: number } {
  const db = dbFromCmd(cmd);
  const epicName = cmd.opts().epic ?? cmd.parent?.opts().epic;
  const epicStore = new EpicStore(db);
  const epic = epicStore.getByName(epicName);
  return { db, epicName, epicId: epic.id };
}

function storyInitCmd(): Command {
  return new Command("init")
    .description("Create a new story within an epic")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .option("--description <desc>", "Story description", "")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const { db, epicName, epicId } = resolveEpic(cmd);
        const store = new StoryStore(db);
        const story = store.create(epicId, name, opts.description);
        writeJSON({
          name,
          epic: epicName,
          path: join(".oraculo", "epics", epicName, "stories", name),
          created: true,
        });
      } catch (err) {
        writeError(err);
      }
    });
}

function storySaveCmd(): Command {
  return new Command("save")
    .description("Save requirements markdown for a story (reads stdin)")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action(async (name: string, opts: any, cmd: Command) => {
      try {
        const { db, epicName, epicId } = resolveEpic(cmd);

        // Upsert DB record
        const store = new StoryStore(db);
        try {
          store.create(epicId, name, "");
        } catch {
          // already exists, ignore
        }

        const content = await Bun.stdin.text();
        const dir = join(".oraculo", "epics", epicName, "stories", name);
        mkdirSync(dir, { recursive: true });
        const path = join(dir, "requirements.md");
        await Bun.write(path, content);

        writeJSON({ name, epic: epicName, path, saved: true });
      } catch (err) {
        writeError(err);
      }
    });
}

function storyGetCmd(): Command {
  return new Command("get")
    .description("Get requirements markdown for a story")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action(async (name: string, opts: any, cmd: Command) => {
      try {
        const { epicName } = resolveEpic(cmd);
        const path = join(
          ".oraculo",
          "epics",
          epicName,
          "stories",
          name,
          "requirements.md",
        );
        const file = Bun.file(path);
        if (!(await file.exists())) {
          writeError(
            new Error(`story "${name}" requirements: file not found`),
          );
          return;
        }
        process.stdout.write(await file.text());
      } catch (err) {
        writeError(err);
      }
    });
}

function storyListCmd(): Command {
  return new Command("list")
    .description("List all stories for an epic")
    .requiredOption("--epic <name>", "Parent epic name")
    .action((opts: any, cmd: Command) => {
      try {
        const { db, epicId } = resolveEpic(cmd);
        const store = new StoryStore(db);
        writeJSON(store.list(epicId));
      } catch (err) {
        writeError(err);
      }
    });
}

function storyUpdateCmd(): Command {
  return new Command("update")
    .description("Update a story's description")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .option("--description <desc>", "New story description", "")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const { db, epicId } = resolveEpic(cmd);
        const store = new StoryStore(db);
        const story = store.update(epicId, name, opts.description);
        writeJSON(story);
      } catch (err) {
        writeError(err);
      }
    });
}

function storyUpdateStatusCmd(): Command {
  return new Command("update-status")
    .description("Update a story's approval status")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .requiredOption(
      "--status <status>",
      "Approval status: none|pending|approved|rejected|needs_revision",
    )
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const status = opts.status as string;
        const validStatuses = ["none", ...APPROVAL_STATUSES];
        if (!validStatuses.includes(status)) {
          writeError(
            new Error(
              `invalid status "${status}": must be one of ${validStatuses.join(", ")}`,
            ),
          );
          return;
        }

        const { db, epicId } = resolveEpic(cmd);
        const store = new StoryStore(db);
        store.updateApprovalStatus(
          epicId,
          name,
          status as ApprovalStatus | "none",
        );
        const story = store.getByName(epicId, name);
        writeJSON(story);
      } catch (err) {
        writeError(err);
      }
    });
}

function storyDeleteCmd(): Command {
  return new Command("delete")
    .description("Delete a story")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const { db, epicName, epicId } = resolveEpic(cmd);
        const store = new StoryStore(db);
        store.delete(epicId, name);
        writeJSON({ name, epic: epicName, deleted: true });
      } catch (err) {
        writeError(err);
      }
    });
}

function storyVersionCmd(): Command {
  return new Command("version")
    .description("Create a new version of story definition (reads stdin)")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action(async (name: string, opts: any, cmd: Command) => {
      try {
        const { db, epicId } = resolveEpic(cmd);

        const storyStore = new StoryStore(db);
        const story = storyStore.getByName(epicId, name);

        const content = await Bun.stdin.text();

        const versionStore = new VersionStore(db);
        const version = versionStore.CreateStoryVersion(story.id, content);

        const approvalStore = new ApprovalStore(db);
        const approval = approvalStore.Create(
          "design",
          epicId,
          story.id,
          content,
        );

        writeJSON({
          version_id: version.id,
          approval_id: approval.id,
        });
      } catch (err) {
        writeError(err);
      }
    });
}

function storyVersionsCmd(): Command {
  return new Command("versions")
    .description("List all versions of a story")
    .argument("<name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const { db, epicId } = resolveEpic(cmd);
        const storyStore = new StoryStore(db);
        const story = storyStore.getByName(epicId, name);

        const versionStore = new VersionStore(db);
        writeJSON(versionStore.ListStoryVersions(story.id));
      } catch (err) {
        writeError(err);
      }
    });
}
